= The quantms.io format
:sectnums:
:toc: left
:doctype: book
//only works on some backends, not HTML
:showcomments:
//use style like Section 1 when referencing within the document.
:xrefstyle: short
:figure-caption: Figure
:pdf-page-size: A4

//GitHub specific settings
ifdef::env-github[]
:tip-caption: :bulb:
:note-caption: :information_source:
:important-caption: :heavy_exclamation_mark:
:caution-caption: :fire:
:warning-caption: :warning:
endif::[]

[[introduction]]
== Introduction

The majority of formats in HUPO-PSI are based on XML format including mzML, mzIdentML making difficult to use them for large-scale, AI model technologies. Also, the previous approach to move away from XML-based approaches, mzTab fails to produce a tab-delimited format that can scale with the size of the data. Here, we aim to formalize and develop a more standardized format that enables better representation of the identification and quantification results but also enables new and novel use cases for proteomics data analysis. The main use cases for the format are:

- Fast and easy visualization of the identification and quantification results.
- Easy integration with other omics data.
- Easy integration with sample metadata.
- AI/ML model development based on identification and quantification results.
- Easy data retrieval for big datasets and large-scale collections of proteomics data.

NOTE: We are not trying to replace the mzTab format, but to provide a new format that enables AI-related use cases.

[[general-data-model]]
== General data model and structure

The quantms.io (.qms) could be seen as a **multiple view** representation of a proteomics data analysis results. Each view of the format can be serialized in different formats depending on the use case. The **data model** defines two main things, the **view** and how the view is **serialized**.

- The **data model view** defines the structure, the fields and properties that will be included in a view for each peptide, psms, feature or protein, or ptms.
- The **data serialization** defines the format in which the view will be serialized and what features of serialization will be supported, for example, compression, indexing, or slicing.

|===
| *view*       | *file class*      | *serialization format* | *definition*
| psm          | psm_file          | _parquet_              | <<psm>>
| feature      | feature_file      | _parquet_              | <<feature>>
| absolute     | absolute_file     | _tsv_                  | <<absolute>>
| differential | differential_file | _tsv_                  | <<differential>>
| sdrf         | sdrf_file         | _tsv_                  | <<sdrf>>
| project      | -                 | _json_                 | <<project>>
|===

TIP: Views can be extended and new views can be added to the format.

The `.qms` folder will contain multiple metadata files that will be used to describe the project, the samples, the data acquisition and the data processing.

[[common-data-structures]]
== Common data structures and formats

We have some concepts that are common for some outputs and would be good to define and explain them here: 

[[peptidoform]]
=== Peptidoform

A peptidoform is a peptide sequence with modifications. For example, the peptide sequence `PEPTIDM` with a modification of `Oxidation` would be `PEPTIDM[Oxidation]`. The peptidoform show be written using the https://github.com/HUPO-PSI/ProForma[Proforma specification ]. This concept is used in the following outputs:

  - <<psm>>
  - <<feature>>
  - <<peptide>>

[[modifications]]
=== Modifications

A modification is a chemical change in the peptide sequence. Modifications can be annotated as part of the Proforma notation inside the peptide or as a separate column. When annotating the modification as a separate column, the format should be as close as possible to the https://github.com/HUPO-PSI/mzTab/tree/master/specification_document-releases/1_0-Proteomics-Release[mzTab format notation]. The modifications will encode the following information on each peptide or psm:

  - Modification name or accession: For example, `Oxidation` or `UNIMOD:35`. Modifications SHOULD be reported using UNIMOD. If a modification is not defined in UNIMOD, a CHEMMOD definition must be used like `CHEMMOD:-18.0913`, where the number is the mass shift in Daltons.
  - Position: The position of the modification in the peptide sequence. Terminal modifications in proteins and peptides MUST be reported with the position set to 0 (N-terminal) or the amino acid length +1 (C-terminal) respectively. For example, `1` or `1,2,3`. 
  - Localization Probability: The probability of the modification being in the reported position. 

Those three properties can be combined in one string as: 

```
{position}({Probabilistic Score:0.9})|{position2}|..-{modification accession or name}
```

For example: 

```
1(Probabilistic Score:0.8)|2(Probabilistic Score:0.9)|3-UNIMOD:35`. 
```

This concept is used in the following outputs:

- <<psm>>
- <<feature>>
- <<peptide>>

[[serialization]]
== Serialization formats

The quantms.io format has different serialization formats for each view. The serialization format defines how the view will be serialized and what features of serialization will be supported, for example, compression, indexing, or slicing. The following serialization formats are supported:

- tsv: Tab-separated values format.
- parquet: Apache Parquet format.
- json: JavaScript Object Notation format.

[[parquet-format]]
=== Parquet format

https://github.com/apache/parquet-format[Parquet] is a columnar storage format that supports nested data. For these large-scale analyses, Parquet has helped its users reduce storage requirements by at least one-third on large datasets, in addition, it greatly improved scan and deserialization time (web use-cases), hence the overall costs. The following table compares the savings as well as the speedup obtained by converting data into Parquet from CSV.

|===
| *Dataset*                            | *Size on Amazon S3* | *Query Run Time* | *Data Scanned*
| Data stored as CSV files             | 1 TB                | 236 seconds      | 1.15 TB
| Data stored in Apache Parquet Format | 130 GB              | 6.78 seconds     | 2.51 GB
|===

[[absolute]]
=== Absolute Quantification View

Absolute quantification is the process of determining the absolute amount of a target protein in a sample. In proteomics, the main computational method to determine the absolute quantification is the intensity-based absolute quantification (iBAQ) method.

==== Absolute Quantification Use cases

- Fast and easy visualization absolute expression (AE) results using iBAQ values.
- Store the AE results of each protein on each sample.
- Provide information about the condition (factor value) of each sample for easy integration.
- Store metadata information about the project, the workflow and the columns in the file.

==== Format

The absolute expression format is a tab-delimited file format that contains the following fields:

-  ``protein`` -> Protein accession or semicolon-separated list of accessions for indistinguishable groups
-  ``sample_accession`` -> Sample accession in the SDRF.
-  ``condition`` -> Condition name
-  ``ibaq`` -> iBAQ value
-  ``ribaq`` -> Relative iBAQ value

Example:

|===
| *protein*    | *sample_accession* | *condition* | *ibaq*  | *ribaq*
| LV861_HUMAN  | Sample-1           | heart       | 1234.1  | 12.34
|===

===== AE Header

By default, the MSstats format does not have any header of metadata. We suggest adding a header to the output for better understanding of the file. By default, MSstats allows comments in the file if the line starts
with ``#``. The quantms output will start with some key value pairs that describe the project, the workflow and also the columns in the file. For

Example:

``#project_accession=PXD000000``

In addition, for each ``Default`` column of the matrix the following information should be added:

   #INFO=<ID=protein, Number=inf, Type=String, Description="Protein Accession">
   #INFO=<ID=sample_accession, Number=1, Type=String, Description="Sample Accession in the SDRF">
   #INFO=<ID=condition, Number=1, Type=String, Description="Value of the factor value">
   #INFO=<ID=ibaq, Number=1, Type=Float, Description="Intensity based absolute quantification">
   #INFO=<ID=ribaq, Number=1, Type=Float, Description="relative iBAQ">

- The ``ID`` is the column name in the matrix, the ``Number`` is the number of values in the column (separated by ``;``), the ``Type`` is the type of the values in the column and the ``Description`` is a description of the column. The number of values in the column can go from 1 to ``inf`` (infinity).
-  Protein groups are written as a list of protein accessions separated by ``;`` (e.g.Â ``P12345;P12346``)

We _RECOMMEND_ including the following properties in the header:

-  project_accession: The project accession in PRIDE Archive
-  project_title: The project title in PRIDE Archive
-  project_description: The project description in PRIDE Archive
-  quantms version: The version of the quantms workflow used to generate the file
-  factor_value: The factor values used in the analysis (e.g.``tissue``)

Please check also the differential expression example for more information: <<differential>>

[[differential]]
=== Differential Expression View

[[psm]]
== Peptide Spectrum Match (PSM) View

Peptide spectrum matches (PSMs) are the results of the identification of peptides in mass spectrometry data. Most of the cases are the results of peptide identified by database search engines on data-dependent acquisition (DDA) experiments.

==== PSM Use cases

-  The PSM table aims to cover detail on PSM level for AI/ML training and other use-cases.
-  Most of the content is similar to mzTab, a PSM would be a peptide identification in a specific msrun file.
-  Store details on PSM level including spectrum mz/intensity for specific use-cases such as AI/ML training.
-  Fast and easy visualization and scanning on PSM level.

==== Format

The PSM view can be found in link:psm.avro[psm.avro].

[[feature]]

Feature

